### ISSUE TITLE
[ban-types] add option to disable a single default ban

### ISSUE THREAD

**Issue by mgcrea** [typescript-eslint/typescript-eslint#2115]
Having upgraded to the latest version I've been hit with the `{}` ban-type when defining React props.

```tsx
type Props = {};

const Button: FunctionComponent<Props> = () => {
  // ...
}
```

I understand that the typing is wrong, but IMO it is a typescript quirk that should be fixed there, not in our code (`{}` should be the type of an empty object) and enforcing this eslint rule will lead to a loss of both productivity (ability to quickly add props to an empty object from a generated user snippet) and readability.

So it would be great at least to allow selective disabling of a ban, because for now, it looks like you have to inline the whole config (minus the excluded rule): eg. maybe by adding an optional flag like:

```
"types": {
        "{}": {
          "disabled": true,
        }
}
```

**Comment by bradzacher** [comment]
Yeah, you're not the first person to ask for this. I think a lot of people disagree with that one because `{}` is pretty ubiquitously "misused".

Should be pretty easy to build out. Probably wouldn't take more than a dozen LOC plus a few tests and docs.  
happy to accept a PR!

Note - I would probably go with a more simple form like passing `false` to disable it, instead of adding a property to the object. Eg
```
{
  "types": {
    "{}": false
  },
  "extendDefaults": true
}
```

**Comment by zetoke** [comment]
@bradleyayers for me personally `{}` makes more sense when I need to have `empty object type`. 

There is a React component example: 
```typescript
interface IState {
  field: string;
}

class MyComponent extends React.Component<{}, IState> {
...
}
```

the default value for props and state is `{}` in React components. And it makes sense since I can catch something like this:
```
public render() {
  const a = this.props.kek;
  console.log(a);
}
```
`Property 'kek' does not exist on type 'Readonly<{}> & Readonly<{ children?: ReactNode; }>'.ts(2339)`

And this error is just muted with `Record<string, unknown>`

**Comment by bradzacher** [comment]
It is misused because people believe it means "an empty object", but it doesn't - it actually means "any non-nullish value".

This is why you can do `const x: {} = 1;`

In particular the type works with react props because typescript has been designed to work with them.

However there is a danger here. If you build utility functions that accept or return your `Props` type, then you have lost some significant type safety because when used outside of the context of react props, the type will accept any non-nullish value.

This is why I say the type is misused - because of the misconception about what it means.

**Comment by fabb** [comment]
The ban-types `{}` rule also errors on generics extending `{}` which is probably not wanted:
```
const deepFreeze = <T extends {}>(o: T): T => {}
```

**Comment by bradzacher** [comment]
It ***definitely*** is wanted.

`{}` means "any non-nullish value".

so with your signature - you might think you're saying "T must be an object", but you're not:

```ts
const deepFreeze = <T extends {}>(o: T): T => { /* freeze somehow */ return o};

deepFreeze(1); // does not error
deepFreeze(false); // does not error
deepFreeze('foo'); // does not error
```
[repl](https://www.typescriptlang.org/play/?ssl=1&ssc=1&pln=6&pc=1#code/MYewdgzgLgBAJgUwQBwGICckC8EwLwwA8AKjAgB5QJhwQwDeAvgHwAUIAXDMQJRel5mDGAHoAVDABmmBDhgQQAWwQALEAHcYYkTExQArujAwQjANwBYAFDXEKDNgSsAjDzOidcEAjpgQsBHR0EHRbJDQZHFZJAEMAGwgENw94b19-MiCQsPtIpwBySRAQfOSRTzSYPwCs0KsgA)

---

An option to disable specific recommended settings has landed. Closing.
